### PR TITLE
dialects: Enable bufferized stencil.apply

### PR DESCRIPTION
**Pull request by PapyChacal**
I.e., make stencil.apply able to express computations on reference-semantic fields through destination operands, instead of only value-semantic through results.

### CODE REVIEW

File: xdsl/dialects/stencil.py [hunk 1]
```diff
@@ -1315,7 +1379,14 @@ def verify_(self) -> None:
...
         apply = cast(ApplyOp, self.parent_op())
-        res_types = [cast(TempType[Attribute], r.type).element_type for r in apply.res]
+        if len(apply.res) > 0:
+            res_types = [
+                cast(TempType[Attribute], r.type).element_type for r in apply.res
+            ]
+        else:
+            res_types = [
+                cast(FieldType[Attribute], o.type).element_type for o in apply.dest
+            ]
```
Comment (n-io) [inline xdsl/dialects/stencil.py:1389]: I think `apply.dest` always has to be present and match, while `apply.res` only has to be there and match in unbufferized mode, no?

File: xdsl/dialects/stencil.py [hunk 2]
```diff
@@ -458,16 +463,27 @@ def print_assign_argument(args: tuple[BlockArgument, SSAValue, Attribute]):
...
         printer.print_list(
             zip(self.region.block.args, self.args, (a.type for a in self.args)),
             print_assign_argument,
         )
         printer.print(") -> (")
-        printer.print_list((r.type for r in self.res), printer.print_attribute)
+        if self.dest:
+            printer.print_list(self.dest, print_destination_operand)
+        else:
+            printer.print_list((r.type for r in self.res), printer.print_attribute)
```
Comment (n-io) [inline xdsl/dialects/stencil.py:480]: Not so keen on this syntax, if I'm understanding correctly that it hides that the bufferized version does not actually have a SSA return value.

Also, can you not have dest+res in the unbufferized mode?

File: xdsl/dialects/stencil.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/dialects/stencil.py:1389]: To simplify a lot other parts, I went with apply having to have either res or dest, never both.

File: xdsl/dialects/stencil.py (hunk 2 above)
Comment (PapyChacal) [reply xdsl/dialects/stencil.py:480]: I went with apply only able to be fully value-semantic (having results in res) or reference-semantics (having destinations in dest), never a mix of both

Comment (PapyChacal) [reply xdsl/dialects/stencil.py:480]: I could clarify that in a docstring I guess

Comment (n-io) [reply xdsl/dialects/stencil.py:480]: My point was something else, that dest is printed after the `->` (which unhelpfully isn't shown in the 4-line snippet above but comes just before)

Comment (n-io) [reply xdsl/dialects/stencil.py:480]: To keep it simple, you could have `-> ( ... )` for results and `dest( .... )` or `outs(....)` for the destination-style passing.

Comment (n-io) [reply xdsl/dialects/stencil.py:480]: You could also consider re-using `print_assign_argument`

Comment (PapyChacal) [reply xdsl/dialects/stencil.py:481]: Done with outs, WDYT?

Comment (n-io) [reply xdsl/dialects/stencil.py:481]: lgtm, thanks for this

**Pull request merged**